### PR TITLE
fixes per external review

### PR DESCRIPTION
**Pull request by ryankurte**

### CODE REVIEW

File: core/src/engine/mod.rs [hunk 1]
```diff
@@ -124,11 +131,11 @@ pub struct Engine<DRV: Driver, RNG: CryptoRngCore = OsRng> {
 /// [`Driver`] trait provides platform support for [`Engine`] instances
 pub trait Driver {
     /// SLIP-0010 derivation for ed25519 keys
-    fn slip10_derive_ed25519(&self, path: &[u32]) -> [u8; 32];
+    fn slip10_derive_ed25519(&self, path: &[u32]) -> Slip10Key;
```
Comment (ryankurte) [inline core/src/engine/mod.rs:134]: prefer `Slip10Key` to `[u8; 32]` for `ZeroizeOnDrop`

File: core/src/engine/mod.rs [hunk 2]
```diff
@@ -926,15 +989,18 @@ impl<DRV: Driver, RNG: CryptoRngCore> Engine<DRV, RNG> {
 
 fn compute_ring_progress(current: usize, ring: usize, total_rings: usize) -> usize {
     // Check to avoid divide by zero
-    if current == 0 && ring == 0 {
+    if (current == 0 && ring == 0) || total_rings == 0 {
```
Comment (ryankurte) [inline core/src/engine/mod.rs:992]: check `total_rings` to avoid divide by zero

File: core/src/engine/ring.rs [hunk 3]
```diff
@@ -98,6 +101,15 @@ impl RingSigner {
         token_id: u64,
         onetime_private_key: Option<TxOnetimeKey>,
     ) -> Result<Self, Error> {
+        // Check ring size and real index are valid
```
Comment (ryankurte) [inline core/src/engine/ring.rs:104]: check `ring_size` and `ring_index` during ring init

File: core/src/engine/summary.rs [hunk 4]
```diff
@@ -71,7 +71,12 @@ impl<const MAX_RECORDS: usize> Summarizer<MAX_RECORDS> {
         num_inputs: usize,
         view_private_key: &RootViewPrivate,
         change_address: &PublicSubaddress,
-    ) -> Self {
+    ) -> Result<Self, Error> {
+        // Check we have some inputs / outputs
+        if num_inputs == 0 || num_outputs == 0 {
```
Comment (ryankurte) [inline core/src/engine/summary.rs:76]: check summary `num_inputs` and `num_outputs` on init

File: fw/src/main.rs [hunk 5]
```diff
@@ -455,12 +455,27 @@ fn handle_apdu<RNG: RngCore + CryptoRng>(
...
+        // Update to identity approver on request
         #[cfg(feature = "ident")]
-        State::Ident(s) if s.is_pending() && !ui.state.is_ident_request() => {
+        State::Ident(IdentState::Pending) if !ui.state.is_ident_request() => {
             ui.state = UiState::IdentRequest(IdentApprover::new());
             render = true;
         }
+        // Show identity state on changes
+        #[cfg(feature = "ident")]
+        State::Ident(IdentState::Approved) => {
```
Comment (ryankurte) [inline fw/src/main.rs:466]: use ident `Approved` and `Rejected` states (`Error` removed)

File: core/src/engine/function.rs [hunk 6]
```diff
@@ -164,7 +164,7 @@ impl Function {
         };
 
         // Initialise summarizer memory
-        unsafe {
+        if let Err(e) = unsafe {
```
Comment (ryankurte) [inline core/src/engine/function.rs:167]: summarizer init now fallible due to bounds checks

**Review by briancorbin — APPROVED** [review]

**Pull request merged**